### PR TITLE
build: Add PRLabeler GitHub Action Workflow

### LINKED ISSUE #3199: Labeling the Pull Request based on the type of changes it is making
## Description

It can be useful to label the PR's in the repository to have a decent visual representation of changes made.

This feature can even be useful when the PR is merged and sometime in future, if we have to trace back to it, it would help us filter out the PR's based on these labels

## Proposed Solution

We can have a GitHub Action Bot, which can label the PR's based on the type of changes it has made.

The labels can somewhat look like the following:
---
image

### PR DESCRIPTION
**Pull request by GandalfGwaihir**
## Description

This Github Action aims to auto label the Pull Requests based on the type of changes made in the repository

Visually the PR's would be labeled similar to the following image:

image

This github action runs on every new PR and  would automatically label the PR's based on the type of changes it has made

## Motivation and Context

Currently, it seems a little difficult visually to figure out which directories/files were updated in a PR, and if we want to navigate back to a merged PR, then this feature can be quite useful as we can filter out PR's based on particular labels on it.

## Review Time Estimate

- [ ] Immediately
- [ ] Within one week
- [x] When possible

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) 
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

 

Closes #3199

### CODE REVIEW

File: .github/workflows/prlabeler.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# This workflow is based on github action official label action v4.
+# This workflow action is triggered on pull request event(on both fork & inside repo)
+# Labels will be applied based on filepath modification in PR.
+# This workflow uses a regex based labeling config file(.github/labeler.yml) to take labeling decision.
+
+name: "PR Labeler"
+on:
+- pull_request_target
```
[the remark this answers is not in the document]
Comment (RohanSasne) [reply .github/workflows/prlabeler.yml:8]: Oh how did i forget the indentation :)

File: .github/workflows/prlabeler.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+      pull-requests: write
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/labeler@v4
+      with:
+        repo-token: "${{ secrets.GITHUB_TOKEN }}"
+        configuration-path: ".github/labeler.yml"
+        sync-labels: false
+        dot: true
```
[the remark this answers is not in the document]
Comment (RohanSasne) [reply .github/workflows/prlabeler.yml:22]: nice observation :) will be more careful about indentation moving forward

File: .github/settings.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#Define the colour of labels over here
```
[the remark this answers is not in the document]
Comment (RohanSasne) [reply .github/settings.yml:1]: Not 100% sure about this but In theory, this shouldn't have any changes to the to the existing labels.

File: .github/labeler.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
+#
+# Add project labels
+#
+
+# Add 'Documentation' label to any changes in the documentation
+
+'Documentation':
+  - documentation/**
```
Comment (infotroph) [inline .github/labeler.yml:8]: Might want to consider adding `book_source/` here too

File: .github/labeler.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  - CONTRIBUTING.md
+  - DEBUGING.md
+  - DEV-INTRO.md
+  - README.md
+  - book_source/**
```
Comment (infotroph) [inline .github/labeler.yml:13]: Just sorting:
```suggestion
  - book_source/**
  - documentation/**
  - CONTRIBUTING.md
  - DEBUGING.md
  - DEV-INTRO.md
  - README.md
```

File: .github/labeler.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+# Add 'Scripts' label to any changes in the scripts directory
+
+'Scripts':
+  - scripts/**
+
+# Add 'Tests' label to any changes in the tests directory
+
+'Tests':
+  - tests/**
```
Comment (infotroph) [inline .github/labeler.yml:53]: While we're at it, might as well tag in-package unit tests too (unlike my point above about CHANGELOG, I think it probably *is* a good thing for PRs that touch a test directory to be labeled 'tests' even if that's not their main point)
```suggestion
  - tests/**
  - '**/tests/**'
  - '!**/tests/Rcheck_reference.log'
```

**Review by infotroph — APPROVED** [review]

**Review by infotroph — APPROVED** [review]

**Pull request merged**